Some fields of DesignDoc should be optional

CouchDB may not return all the fields currently needed to create a DesignDoc instance, hence calls to `design.getDesignDocument` may fail.

```
[error] Exception in thread "main" spray.json.DeserializationException: Object is missing required member 'updates'
[error] 	at spray.json.package$.deserializationError(package.scala:23)
[error] 	at spray.json.ProductFormats$class.fromField(ProductFormats.scala:60)
[error] 	at gnieh.sohva.SohvaProtocol$.fromField(SohvaProtocol.scala:282)
[error] 	at spray.json.ProductFormatsInstances$$anon$9.read(ProductFormatsInstances.scala:262)
[error] 	at spray.json.ProductFormatsInstances$$anon$9.read(ProductFormatsInstances.scala:242)
[error] 	at gnieh.sohva.SohvaProtocol$$anon$1.read(SohvaProtocol.scala:59)
[error] 	at gnieh.sohva.SohvaProtocol$$anon$1.read(SohvaProtocol.scala:59)
[error] 	at spray.json.JsValue.convertTo(JsValue.scala:31)
[error] 	at gnieh.sohva.Design.gnieh$sohva$Design$$designDoc(Design.scala:345)
[error] 	at gnieh.sohva.Design$$anonfun$getDesignDocument$1$$anonfun$apply$2.apply(Design.scala:70)
[error] 	at gnieh.sohva.Design$$anonfun$getDesignDocument$1$$anonfun$apply$2.apply(Design.scala:70)
[error] 	at scala.Option.map(Option.scala:146)
[error] 	at gnieh.sohva.Design$$anonfun$getDesignDocument$1.apply(Design.scala:70)
[error] 	at gnieh.sohva.Design$$anonfun$getDesignDocument$1.apply(Design.scala:68)
[error] 	at scala.util.Success$$anonfun$map$1.apply(Try.scala:237)
[error] 	at scala.util.Try$.apply(Try.scala:192)
[error] 	at scala.util.Success.map(Try.scala:237)
[error] 	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:237)
[error] 	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:237)
[error] 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
[error] 	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
[error] 	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply$mcV$sp(BatchingExecutor.scala:91)
[error] 	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:91)
[error] 	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:91)
[error] 	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
[error] 	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:90)
[error] 	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:39)
[error] 	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:409)
[error] 	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
[error] 	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
[error] 	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
[error] 	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
[error] Caused by: java.util.NoSuchElementException: key not found: updates
[error] 	at scala.collection.MapLike$class.default(MapLike.scala:228)
[error] 	at scala.collection.AbstractMap.default(Map.scala:59)
[error] 	at scala.collection.MapLike$class.apply(MapLike.scala:141)
[error] 	at scala.collection.AbstractMap.apply(Map.scala:59)
[error] 	at spray.json.ProductFormats$class.fromField(ProductFormats.scala:57)
[error] 	... 30 more
```

--- a/src/main/scala/gnieh/sohva/SohvaProtocol.scala
+++ b/src/main/scala/gnieh/sohva/SohvaProtocol.scala
@@ -56,7 +56,46 @@ trait SohvaProtocol extends DefaultJsonProtocol with MangoProtocol with CouchFor
 
   implicit val attachmentFormat = jsonFormat7(Attachment)
 
-  implicit val designDocFormat = couchFormat[DesignDoc]
+  implicit val designDocFormat = new CouchFormat[DesignDoc] {
+    def _id(d: DesignDoc) = d._id
+    def _rev(d: DesignDoc) = d._rev
+    def withRev(d: DesignDoc, r: Option[String]) = d.withRev(r)
+
+    def read(json: JsValue): DesignDoc = json match {
+      case JsObject(fields) if fields.contains("_id") =>
+        val id = fields("_id").convertTo[String]
+        val rev = fields("_rev").convertTo[Option[String]]
+        val language = fields.get("language").map(_.convertTo[String]).getOrElse("javascript")
+        val views = fields.get("views").map(_.convertTo[Map[String, ViewDoc]]).getOrElse(Map())
+        val validate_doc_update = fields.get("validate_doc_update").map(_.convertTo[String])
+        val updates = fields.get("updates").map(_.convertTo[Map[String, String]]).getOrElse(Map())
+        val filters = fields.get("filters").map(_.convertTo[Map[String, String]]).getOrElse(Map())
+        val shows = fields.get("shows").map(_.convertTo[Map[String, String]]).getOrElse(Map())
+        val lists = fields.get("lists").map(_.convertTo[Map[String, String]]).getOrElse(Map())
+        val rewrites = fields.get("rewrites").map(_.convertTo[List[RewriteRule]]).getOrElse(Nil)
+        DesignDoc(id, language, views, validate_doc_update, updates, filters, shows, lists, rewrites).withRev(rev)
+      case _ =>
+        deserializationError(f"Malfromed design document $json")
+    }
+
+    def write(d: DesignDoc): JsObject = {
+      val DesignDoc(id, language, views, validate_doc_update, updates, filters, shows, lists, rewrites) = d
+      val fields = List(
+        Some("_id" -> id.toJson),
+        d._rev.map("_rev" -> _.toJson),
+        Some("language" -> language.toJson),
+        if (views.isEmpty) None else Some("views" -> views.toJson),
+        validate_doc_update.map("validate_doc_update" -> _.toJson),
+        if (updates.isEmpty) None else Some("updates" -> updates.toJson),
+        if (views.isEmpty) None else Some("filters" -> filters.toJson),
+        if (views.isEmpty) None else Some("shows" -> shows.toJson),
+        if (views.isEmpty) None else Some("lists" -> lists.toJson),
+        if (views.isEmpty) None else Some("rewrites" -> rewrites.toJson))
+        .flatten
+        .toMap
+      JsObject(fields)
+    }
+  }
 
   implicit object DateFormat extends JsonFormat[Date] {
 

